lib/Dialect/Arc/Transforms/MakeTables.cpp

```diff
@@ -0,0 +1,193 @@
+  void runOnOperation() override;
+  void runOnDefine();
+  DefineOp defineOp;
+};
+} // namespace
+
+static inline uint32_t bitsMask(uint32_t nbits) {
+  if (nbits == 32)
+    return ~0;
+  return (1 << nbits) - 1;
+}
+
+static inline uint32_t bitsGet(uint32_t x, uint32_t lb, uint32_t ub) {
+  return (x >> lb) & bitsMask(ub - lb + 1);
+}
+
+void MakeTablesPass::runOnOperation() {
+  auto module = getOperation();
+  for (auto op : llvm::make_early_inc_range(module.getOps<DefineOp>())) {
+    defineOp = op;
+    runOnDefine();
+  }
+}
+
+void MakeTablesPass::runOnDefine() {
+  LLVM_DEBUG(llvm::dbgs() << "Making lookup tables in `" << defineOp.getName()
+                          << "`\n");
+
+  unsigned inbits = 0;
+  unsigned outbits = 0;
+  unsigned opcount = defineOp.getBodyBlock().getOperations().size();
+
+  // inputs
+  bool noninteger = false;
+  auto types = defineOp.getArgumentTypes();
+  for (auto &t : types) {
+    if (auto i = t.dyn_cast<IntegerType>()) {
+      inbits += i.getWidth();
+    } else {
+      noninteger = true;
+    }
+  }
+  if (noninteger) {
+    // only make lookup tables if all inputs are integers
+    return;
+  }
+  for (auto op :
+       llvm::make_early_inc_range(defineOp.getOps<hw::ConstantOp>())) {
+    (void)op;
+    opcount--; // don't count constant ops
+  }
+  // outputs
+  arc::OutputOp outputOp;
+  for (auto op : llvm::make_early_inc_range(defineOp.getOps<arc::OutputOp>())) {
+    outputOp = op;
+    opcount--; // don't count output ops
+    auto types = op.getOperandTypes();
+    for (auto t : types)
+      if (auto i = t.dyn_cast<IntegerType>())
+        outbits += i.getWidth();
+  }
+  LLVM_DEBUG(llvm::dbgs() << "lookup table analysis: inbits: " << inbits
+                          << ", outbits: " << outbits
+                          << ", num ops: " << opcount << "\n");
+
+  if (inbits > 26 || outbits > 26) {
+    LLVM_DEBUG(llvm::dbgs() << "no table created: table is too large"
+                            << "\n");
+    return;
+  }
+
+  unsigned insize = 1U << inbits;
+  const unsigned tableBytes = insize * outbits / 8;
+
+  if (opcount < tableMinOpCount) {
+    LLVM_DEBUG(llvm::dbgs() << "no table created: too few ops involved"
+                            << "\n");
+    return;
+  }
+  if (tableBytes > tableMaxBytes) {
+    LLVM_DEBUG(llvm::dbgs() << "no table created: table is too large"
+                            << "\n");
+    return;
+  }
+  if (!outbits || !inbits) {
+    LLVM_DEBUG(llvm::dbgs() << "no table created: no outputs or no inputs"
+                            << "\n");
+    return;
+  }
+
+  LLVM_DEBUG(llvm::dbgs() << "creating table of size " << tableBytes << "\n");
+
+  SmallVector<Operation *, 64> origBody;
+  for (auto &op : defineOp.getBodyBlock()) {
+    if (isa<arc::OutputOp>(op)) {
+      continue;
+    }
+    origBody.push_back(&op);
+  }
+
+  auto b = ImplicitLocOpBuilder::atBlockBegin(defineOp.getLoc(),
```

Comment:
The code from here to the end of the function could be replaced by a few lines to create a `arc.lut` operation (provided that the PR with that op and the LowerLUT pass lands first). But we can also do that in a follow-up PR, because one annoyance will be that the LUTOp does not support multiple outputs, so it requires insertion of a a concat and some extract operations. The LowerLUT pass contains a revised version of this code that should offer considerably higher compilation performance.
BTW: This part uses the part of the folders where all inputs are constant, so factoring out these parts as discussed in the last simulation themed weekly meeting to not use Attributes would lead to nice speedups here (because if there are many tables inserted, this pass will likely be slow).

Reply:
Yeah that sounds like a great plan. The LowerLUT lands later in the PR stack. Let's do that follow-up PR you've suggested!